paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/PrimaryKeyLRULookupTable.java

```diff
@@ -0,0 +1,226 @@
+                                options.path(),
+                                partitionType,
+                                options.partitionDefaultName(),
+                                options.fileFormat().getFormatIdentifier()),
+                        PrimaryKeyTableUtils.PrimaryKeyFieldsExtractor.EXTRACTOR,
+                        options);
+        readerFactoryBuilder.withValueProjection(Projection.of(projection).toNestedIndexes());
+        this.projectedValueType = readerFactoryBuilder.projectedValueType();
+        this.keyComparatorSupplier = new KeyComparatorSupplier(keyType);
+        this.hashLookupStoreFactory =
+                new HashLookupStoreFactory(
+                        cacheManager,
+                        options.toConfiguration().get(CoreOptions.LOOKUP_HASH_LOAD_FACTOR));
+    }
+
+    @Override
+    public List<InternalRow> get(InternalRow key) throws IOException {
+        LookupLevels lookupLevels = getLookupLevels(key);
+        if (lookupLevels == null) {
+            return Collections.emptyList();
+        } else {
+            // lookup start from level 1.
+            KeyValue kv = lookupLevels.lookup(key, 1);
+            if (kv == null || kv.valueKind().isRetract()) {
+                return Collections.emptyList();
+            } else {
+                return Collections.singletonList(kv.value());
+            }
+        }
+    }
+
+    private LookupLevels getLookupLevels(InternalRow key) {
+        extractor.setRecord(key);
+        int bucket = extractor.bucket();
+
+        if (tableView.isEmpty()) {
+            return null;
+        }
+        // Non partitioned table
+        Map<Integer, LookupLevels> buckets = tableView.entrySet().iterator().next().getValue();
+        return buckets.get(bucket);
+    }
+
+    @Override
+    public void refresh(List<Split> splits) {
+        for (Split split : splits) {
+            if (!(split instanceof DataSplit)) {
+                throw new IllegalArgumentException("Unsupported split: " + split.getClass());
+            }
+            BinaryRow partition = ((DataSplit) split).partition();
+            int bucket = ((DataSplit) split).bucket();
+            List<DataFileMeta> before = ((DataSplit) split).beforeFiles();
+            List<DataFileMeta> after = ((DataSplit) split).dataFiles();
+
+            Map<Integer, LookupLevels> buckets =
+                    tableView.computeIfAbsent(partition, k -> new HashMap<>());
+            LookupLevels lookupLevels = buckets.get(bucket);
+
+            if (lookupLevels == null) {
+                Preconditions.checkArgument(
+                        before.isEmpty(), "The before file should be empty for the initial phase.");
+                Levels levels = new Levels(keyComparatorSupplier.get(), after, options.numLevels());
+
+                KeyValueFileReaderFactory factory = readerFactoryBuilder.build(partition, bucket);
+
+                lookupLevels =
+                        new LookupLevels(
```

Comment:
Can you add an interface for lookup? For example:
1. Add `FileStoreTable.newQuery(@Nullable int[] projection)`.
2. `PrimaryKeyFileStoreTable` override `newQuery`.

```
public interface TableQuery extends Closeable {

    void refreshFiles(
            BinaryRow partition,
            int bucket,
            List<DataFileMeta> beforeFiles,
            List<DataFileMeta> dataFiles);

    @Nullable
    InternalRow lookup(BinaryRow partition, int bucket, InternalRow key) throws IOException;
}
```

This can be a separate PR.

Reply:
Updated in the latest commit.